### PR TITLE
Enable use of internal tensors in Sequential

### PR DESCRIPTION
**Pull request by janekb04**
# Description

Enable use of internal tensors (ex. `Float8TensorBase`) instead of regular quantized tensor (ex. `Float8Tensor`) in `te.Sequential`. Essentially, this removes calls to `Float8Tensor.__new__` and to `detach` (see visual comparison of forward pass below).

Expected performance gain: E2E FP8 GPT encoder Transformer layer implemented using Sequential: +15% over `main`.

## Visual comparison

image

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

1. (39ea63da05c23565201e2270626afbcc8672155a) Replace free function `is_float8_tensor` with `is_quantized_tensor` in `_common.py` and use it throughout the `ops` codebase to check if a tensor is a (possibly internal) quantized tensor
2. (9e5acd94af7ceaa3a94598474bbebac162b5dcdd) Change interface of `fuser_forward` and `op_forward` to no longer take preceding and following ops and instead take the following op's input quantizer and preceding op's input gradient's quantizer directly
3. (c226033f5c292ee07b952716e40f8148eefe7736) Remove use of `detach` on input tensors saved for backward (enabled by not passing `prev_op` to backward)
4. (69e8cd564b8fbd403c03d414f337e29b5c70d99b) Handle saving internal tensors in `_OperationFuserAutogradFunction` using `prepare_for_saving` and `restore_from_saved`
5. (563d8b9cced7ab19ffe4f2be3e29325d67e046e3) Enable use of internal tensors in `BasicLinear` quantizers and fix issues resulting from internal tensors not having methods that regular tensors have

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Performance data

Below is reported the running time of a GPT encoder transformer layer (averaged over 10k runs) using [my benchmark script](https://github.com/janekb04/transformer-engine-benchmark-sequential). Here, Fused refers to an implementation using the fused modules `LayerNormLinear` and `LayerNormMLP`, Sequential refers to an implementation implementing those two modules with `te.ops.Sequential`, and Builtin refers to using `te.TransformerLayer`.

### B100 results

|Commit|Fused BF16|Fused FP8|Sequential BF16|Sequential FP8|Builtin BF16|Builtin FP8|
|-|-|-|-|-|-|-|
|cc0cb35d4cb516669e1f166add37ccdc30f71488|1.60ms|1.90ms|1.55ms|2.39ms|2.21ms|2.61ms|
|39ea63da05c23565201e2270626afbcc8672155a|1.60ms|1.90ms|1.55ms|2.39ms|2.21ms|2.60ms|
|9e5acd94af7ceaa3a94598474bbebac162b5dcdd|1.62ms|1.92ms|1.56ms|2.39ms|2.21ms|2.60ms|
|c226033f5c292ee07b952716e40f8148eefe7736|1.57ms|1.89ms|1.54ms|2.34ms|2.22ms|2.58ms|
|69e8cd564b8fbd403c03d414f337e29b5c70d99b|1.58ms|1.92ms|1.55ms|2.36ms|2.22ms|2.59ms|
|563d8b9cced7ab19ffe4f2be3e29325d67e046e3|1.58ms|1.93ms|1.55ms|2.05ms|2.22ms|2.61ms|

### H100 results

|Commit|Fused BF16|Fused FP8|Sequential BF16|Sequential FP8|Builtin BF16|Builtin FP8|
|-|-|-|-|-|-|-|
|cc0cb35d4cb516669e1f166add37ccdc30f71488|1.99ms|2.56ms|2.08ms|3.33ms|2.83ms|3.42ms|
|39ea63da05c23565201e2270626afbcc8672155a|2.02ms|2.61ms|2.09ms|3.35ms|2.78ms|3.40ms|
|9e5acd94af7ceaa3a94598474bbebac162b5dcdd|2.11ms|2.67ms|2.11ms|3.33ms|2.79ms|3.31ms|
|c226033f5c292ee07b952716e40f8148eefe7736|2.07ms|2.64ms|2.09ms|3.27ms|2.80ms|3.41ms|
|69e8cd564b8fbd403c03d414f337e29b5c70d99b|2.10ms|2.65ms|2.10ms|3.27ms|2.81ms|3.41ms|
|563d8b9cced7ab19ffe4f2be3e29325d67e046e3|2.09ms|2.64ms|2.09ms|2.85ms|2.83ms|3.41ms|

### CODE REVIEW

File: transformer_engine/pytorch/ops/_common.py [hunk 1]
```diff
@@ -17,16 +17,16 @@
 from ..utils import canonicalize_dtype
 
 
-def is_float8_tensor(tensor: Any) -> bool:
-    """Check if object is a `Float8Tensor`"""
-    return isinstance(tensor, Float8Tensor)
+def is_quantized_tensor(tensor: torch.Tensor | QuantizedTensorBase):
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/_common.py]:
```suggestion
def is_quantized_tensor(tensor: torch.Tensor | QuantizedTensorBase) -> bool:
```

File: transformer_engine/pytorch/ops/basic/identity.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
     BasicOperation,
     OperationContext,
 )
+from transformer_engine.pytorch.tensor import Quantizer
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/identity.py]: To force local imports:
```suggestion
from ...tensor import Quantizer
```

File: transformer_engine/pytorch/ops/basic/make_extra_output.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
     BasicOperation,
     OperationContext,
 )
+from transformer_engine.pytorch.tensor import Quantizer
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/make_extra_output.py]:
```suggestion
from ...tensor import Quantizer
```

File: transformer_engine/pytorch/ops/basic/reshape.py [hunk 4]
```diff
@@ -14,6 +14,7 @@
     BasicOperation,
     OperationContext,
 )
+from transformer_engine.pytorch.tensor import Quantizer
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/reshape.py]:
```suggestion
from ...tensor import Quantizer
```

File: transformer_engine/pytorch/ops/basic/basic_linear.py [hunk 5]
```diff
@@ -560,18 +563,14 @@ def _functional_forward(
...
                 w.update_usage(rowwise_usage=False, columnwise_usage=True)
         else:
             w = None
 
         # Prepare input tensor for backward pass
         if weight_requires_grad:
-            if x_local is input:
-                # PyTorch autograd produces esoteric errors if we
-                # cache input tensor directly.
-                x_local = x_local.detach()
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/basic_linear.py:573]: Pinning this change in case we see this error again in the future. If it shows up, we might need to do something like:
```python
if x_local is input and isinstance(x_local, torch.Tensor):
    x_local = x_local.detach()
```

File: transformer_engine/pytorch/ops/fused/forward_linear_bias_activation.py [hunk 6]
```diff
@@ -13,11 +13,11 @@
 from transformer_engine.pytorch.fp8 import FP8GlobalStateManager
 from transformer_engine.pytorch.ops.basic import BasicLinear, Bias
 from transformer_engine.pytorch.ops.op import (
-    BasicOperation,
     FusedOperation,
     FusibleOperation,
     OperationContext,
 )
+from transformer_engine.pytorch.tensor import Quantizer
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/fused/forward_linear_bias_activation.py]:
```suggestion
from ...tensor import Quantizer
```

File: transformer_engine/pytorch/ops/basic/basic_linear.py [hunk 7]
```diff
@@ -886,12 +885,13 @@ def op_forward(
...
         input_: torch.Tensor,
-        prev_op: Optional[BasicOperation] = None,
-        next_op: Optional[BasicOperation] = None,
+        prev_op_grad_input_quantizer: Optional[Quantizer],
+        next_op_input_quantizer: Optional[Quantizer],
+        is_first_op: bool,
     ) -> torch.Tensor:
 
         # Check which grads are required
-        input_requires_grad = ctx.requires_grad and input_.requires_grad
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/basic/basic_linear.py:894]: We want to handle the case where `input_requires_grad=False` and `weight_requires_grad=True` so that we can skip one of the backward GEMMs. I think we'll need to expand the `requires_grad` logic in the fuser with something like:
```python
for idx in basic_op_idxs:
    basic_op_ctxs[idx].requires_grad = ...
    basic_op_ctxs[idx].input_requires_grad = input_.requires_grad if idx == 0 else basic_op_ctxs[idx-1].requires_grad
```

File: transformer_engine/pytorch/ops/basic/basic_linear.py (hunk 5 above)
Comment (janekb04) [reply transformer_engine/pytorch/ops/basic/basic_linear.py:573]: I'm afraid the autograd error occurred even when using internal tensors, so this wouldn't fix the issue, if it occurred again. As the internal tensors don't implement `detach`, there was no way to fix it (without first implementing `detach` for them).

File: transformer_engine/pytorch/ops/basic/basic_linear.py (hunk 7 above)
Comment (janekb04) [reply transformer_engine/pytorch/ops/basic/basic_linear.py:894]: I removed the check for `input_.requires_grad` for two reasons: 1. internal tensors don't implement it 2. as far as I could see, it was guaranteed that `input_.requires_grad == ctx.requires_grad`, as `input_.requires_grad` was explicitly adjusted to match `ctx.requires_grad` in `fuser.py::forward:120-124`.

Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/basic_linear.py:894]: I see, better to keep this PR functionally equivalent to the existing code then. We'll pin this as a future optimization.

**Review by timmoon10 — APPROVED** [review]
LGTM
